go.mod

```diff
@@ -2,13 +2,16 @@ module github.com/open-telemetry/opentelemetry-collector-contrib
 
 go 1.12
 
+replace github.com/open-telemetry/opentelemetry-collector-contrib/exporter/azuremonitorexporter v0.0.0 => ./exporter/azuremonitorexporter
```

Comment:
Why is this replace needed?

Reply:
Following the examples for the other additions in this repo:
https://github.com/pcwiese/opentelemetry-collector-contrib/blob/2254e5b860a679f1bde85eca0bfc6519a25cd3f6/cmd/otelcontribcol/components.go#L24

They are referenced via modules, but this is new, and there is no module for it yet, hence this mapping. This mapping appeared to be required to get it to build at all using the same pattern. Is there another way to do this? I assumed once this gets merged, the module will be available for reference just like the others.

Reply:
I am not a Go modules expert, but I'd expect it to build without the need replace because this is the local repository. I am be wrong. Can you post the the build error you get?
(Either way if this is needed temporarily we will need to remove it after merging).

Reply:
If I remove this mapping, the build fails with:
build github.com/open-telemetry/opentelemetry-collector-contrib/cmd/otelcontribcol: cannot load github.com/open-telemetry/opentelemetry-collector-contrib/exporter/azuremonitorexporter: no matching versions for query "latest"

I would actually argue that the default build should always use this replacement mapping syntax for the receivers and exporters in this repo. If you want something other than the tip of the tree, switch to a specific version and replace the mapping.

Reply:
Do you see this failure locally on your machine or in CI of this repo?
Something looks wrong to me. This should not be happening. 

`exporter/azuremonitorexporter` is a local package, why would the build need to query its version and load it?

@owais you created the build, can you please check this? Why do we need components that are local in this repo to be listed explicitly in top-level `go.mod` file? This doesn't look right to me either.

Reply:
failure is local for me.

Reply:
I did not make the decision for each component to be it's own module so I might not have the whole picture but I think it was mainly done to allow the community to pick and choose components they want and create custom builds. If we don't make each component a Go module on it's own then any other will have to pull in all of opentelemetry-collector-contrib module (the whole repo) as a dependency even if they needed just one component. Another downside is that people cannot use different versions of different components. Let's say someone wanted to use exporterA from version v1.1.0 but at the same time wanted to use receiverA from version v1.5.0, they won't be able to do it as their build would pull in the whole contrib repo as a single dependency. 

This is not much of a problem now but hopefully as this repo grows to multi-dozen or even hundreds of components, I think it'll become too rigid to ship all of them as a single package. 

---

To answer the specific question as to why the replace is needed, it is needed only for the first PR until the component is accepted into master. Since the component is itself a Go module, Go tries to look for it on Github given it's import path and doesn't find it since it is not published yet. Once the PR is merged, the replace directive can be removed as Go will then find the module on Github. That also means that the contrib build will use the fixed version of the module going forward. Every time a module/component is updated, root go.mod will have to be updated as well to use the new version. This is a bit of additional work when initially building modules but it gives us a lot more flexibility in how we ship and consume the contrib packages.  Again, since I did not make the original decision, I might be missing some other reasons.

Reply:
@owais make sense to me. Thanks for the detailed explanation. All good points. Let's keep the component modules.

Reply:
@owais assessment here is correct. In the other PR that I'm adding the SignalFx exporter I opted to not merge the `replace` but I will do a smaller follow up PR adding it to the executable.